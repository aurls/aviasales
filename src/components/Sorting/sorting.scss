@import '../../constants/styleConstants.scss';

.sorting {
  display: flex;
  justify-content: space-between;
  background: white;
  margin-bottom: 20px;
}

.sorting__item {
  display: flex;
  flex: 1;
  justify-content: center;
  align-items: center;
  width: auto;
  height: 50px;
  border: 1px solid #dfe5ec;
  border-right: 0;
  border-radius: 5px;
  cursor: pointer;
  font-size: 12px;
  line-height: 20px;
  font-weight: 600;
  letter-spacing: .5px;
  text-transform: uppercase;
  color: #4a4a4a;

  &:hover {
    background: $lightPrimaryColor;
    transition: all $standardTransition;
  }

  &:first-child {
    border-radius: 5px 0 0 5px;
  }

  &:last-child {
    border-radius: 0 5px 5px 0;
    border-right: 1px solid #dfe5ec;
  }

  &.is-active {
    border-color: $primaryColor;
    background: $primaryColor;
    color: white;
  }
}
