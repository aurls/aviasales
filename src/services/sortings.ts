import Types from '../types';

const BY_PRICE_ASC: Types.Sorting = {
  type: 'BY_PRICE_ASC',
  title: 'Самый дешёвый',
  fn: (prevPrice, nextPrice) => prevPrice - nextPrice
};

const BY_SPEED_ASC: Types.Sorting = {
  type: 'BY_SPEED_ASC',
  title: 'Самый быстрый',
  fn: (prevSpeed, nextSpeed) => prevSpeed - nextSpeed
};

export default {
  BY_PRICE_ASC,
  BY_SPEED_ASC
};
