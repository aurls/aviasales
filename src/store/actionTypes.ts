export default ({
  FETCH_TICKETS_REQUEST: 'FETCH_TICKETS_REQUEST',
  FETCH_TICKETS_SUCCESS: 'FETCH_TICKETS_SUCCESS',
  FETCH_TICKETS_FAILURE: 'FETCH_TICKETS_FAILURE',
  SET_FETCHING: 'SET_FETCHING',
  SET_FILTER: 'SET_FILTER',
  SET_SORTING: 'SET_SORTING'
} as const);
