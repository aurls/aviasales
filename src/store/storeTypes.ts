import Types from '../types';
import actionTypes from './actionTypes';

export namespace Actions {
  export type FetchTicketsRequest = {
    type: typeof actionTypes.FETCH_TICKETS_REQUEST
  }

  export type FetchTicketsSuccess = {
    type: typeof actionTypes.FETCH_TICKETS_SUCCESS,
    payload: Types.Tickets
  }

  export type FetchTicketsFailure = {
    type: typeof actionTypes.FETCH_TICKETS_FAILURE
  }

  export type SetFilter = {
    type: typeof actionTypes.SET_FILTER,
    payload: Types.Filter
  }

  export type SetSorting = {
    type: typeof actionTypes.SET_SORTING,
    payload: string
  }

  export type All =
    | FetchTicketsRequest
    | FetchTicketsSuccess
    | FetchTicketsFailure
    | SetFilter
    | SetSorting
}

export type State = {
  tickets: Types.Tickets,
  isFetching: boolean,
  hasError: boolean,
  filter: Types.Filter,
  sorting: string
}
