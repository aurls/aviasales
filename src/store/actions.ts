import Types from '../types';
import { Actions } from './storeTypes';
import actionTypes from './actionTypes';

const fetchTicketsRequest = (): Actions.FetchTicketsRequest => ({
  type: actionTypes.FETCH_TICKETS_REQUEST
});

const fetchTicketsSuccess = (tickets: Types.Tickets): Actions.FetchTicketsSuccess => ({
  type: actionTypes.FETCH_TICKETS_SUCCESS,
  payload: tickets
});

const fetchTicketsFailure = (): Actions.FetchTicketsFailure => ({
  type: actionTypes.FETCH_TICKETS_FAILURE
});

const setFilter = (filter: Types.Filter): Actions.SetFilter => ({
  type: actionTypes.SET_FILTER,
  payload: filter
});

const setSorting = (sorting: string): Actions.SetSorting => ({
  type: actionTypes.SET_SORTING,
  payload: sorting
});

export default {
  fetchTicketsRequest,
  fetchTicketsSuccess,
  fetchTicketsFailure,
  setFilter,
  setSorting
};
