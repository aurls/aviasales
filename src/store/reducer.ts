import Types from '../types';
import actionTypes from './actionTypes';
import { Actions, State } from './storeTypes';
import filters from '../services/filters';
import sortings from '../services/sortings';

const initialState: State = {
  tickets: {},
  isFetching: true,
  hasError: false,
  filters: [filters.ALL],
  sorting: sortings.BY_PRICE_ASC
};

const reducer = (state: State = initialState, action: Actions.All): State => {
  switch (action.type) {
    case actionTypes.FETCH_TICKETS_REQUEST:
      return fetchTicketsRequest(state);

    case actionTypes.FETCH_TICKETS_SUCCESS:
      return fetchTicketsSuccess(state, action.payload);

    case actionTypes.FETCH_TICKETS_FAILURE:
      return fetchTicketsFailure(state);

    case actionTypes.SET_FILTER:
      return setFilter(state, action.payload);

    case actionTypes.SET_SORTING:
      return setSorting(state, action.payload);

    default:
      return state;
  }
};

function fetchTicketsRequest (state: State): State {
  return {
    ...state,
    isFetching: true,
    hasError: false
  };
}

function fetchTicketsSuccess (state: State, newTickets: Types.Tickets): State {
  return {
    ...state,
    tickets: {
      ...state.tickets,
      ...newTickets
    },
    isFetching: false,
    hasError: false
  };
}

function fetchTicketsFailure (state: State): State {
  return {
    ...state,
    isFetching: false,
    hasError: true
  };
}

function setFilter (state: State, filter: Types.Filter): State {
  const updateFilters = (filters: Types.Filter[]) => ({
    ...state,
    filters
  });
  const currentFilters: Types.Filter[] = [...state.filters];

  if (filter === filters.ALL) {
    if (currentFilters.includes(filters.ALL)) {
      return updateFilters([]);
    }
    return updateFilters([filters.ALL]);
  }

  if (currentFilters.includes(filter)) {
    return updateFilters(
      [...currentFilters.filter((i) => i !== filter)]);
  }

  const filtersCount = Object.values(filters).length;
  const currentFiltersCount = Object.values(currentFilters).length;

  if (filtersCount - currentFiltersCount > 2) {
    return updateFilters(
      [...currentFilters.filter((i) => i !== filters.ALL), filter]);
  }

  return updateFilters([filters.ALL]);
}

function setSorting (state: State, sorting: Types.Sorting): State {
  return {
    ...state,
    sorting
  };
}

export default reducer;
