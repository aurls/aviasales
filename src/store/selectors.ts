import { createSelector } from 'reselect';
import Types from '../types';
import { State } from './storeTypes';
import filters from '../services/filters';
import filterTickets from '../services/filterTickets';
import sortTickets from '../services/sortTickets';

const getAllTickets = (state: State): Types.Tickets => state.tickets;
const getFilter = (state: State): Types.Filter => state.filter;
const getSorting = (state: State): string => state.sorting;
const getIsFetching = (state: State): boolean => state.isFetching;
const getHasError = (state: State): boolean => state.hasError;

const getTickets = createSelector(
  getAllTickets,
  getFilter,
  getSorting,
  (tickets: Types.Tickets, filter: Types.Filter, sorting: string): Types.Tickets => {
    return tickets;
  }
);

export default {
  getTickets,
  getFilter,
  getSorting,
  getIsFetching,
  getHasError
};
