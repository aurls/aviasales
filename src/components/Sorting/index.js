import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import * as sorting from '../../constants/sorting';
import { sorting as sortingActions } from '../../actions';
import './sorting.scss';

const Sorting = ({ appliedSorting, setSorting }) => {
  const onChangeSorting = (event) => {
    setSorting(event.target.id);
  };

  return (
    <ul className="sorting">
      {
        Object.values(sorting).map((item) => {
          let itemStyle = 'sorting__item';
          if (appliedSorting.includes(item.type)) itemStyle += ' is-active';
          return (
            <li key={item.type}
              className={itemStyle}
              id={item.type}
              onClick={onChangeSorting}>
              {item.name}
            </li>);
        })
      }
    </ul>
  );
};

const mapStateToProps = (state) => {
  return {
    appliedSorting: state.appliedSorting
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    setSorting: bindActionCreators(sortingActions.setSorting, dispatch)
  };
};

Sorting.propTypes = {
  appliedSorting: PropTypes.array.isRequired,
  setSorting: PropTypes.func.isRequired
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Sorting);
