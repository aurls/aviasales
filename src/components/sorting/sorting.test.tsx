import React from 'react';
import Enzyme from 'enzyme';
import Sorting from './index';
import { initialState } from '../../store/reducer';
import sortings from '../../services/sortings';

describe('Sorting component', () => {
  const setSorting = jest.fn();
  let component: Enzyme.ShallowWrapper;

  beforeEach(() => {
    component = Enzyme.shallow(
      <Sorting
        currentSorting={initialState.sorting}
        setSorting={setSorting} />
    );
  });

  it('Default state', () => {
    expect(component).toMatchSnapshot();
  });

  it('Sorting changed', () => {
    const selector = '.sorting__item.is-active';
    component.find(selector).simulate('click');

    expect(setSorting).toHaveBeenCalledWith(sortings.BY_PRICE_ASC);
  });
});
