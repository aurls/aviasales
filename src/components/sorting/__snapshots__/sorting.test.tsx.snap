// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sorting component Default state 1`] = `
<ul
  className="sorting"
>
  <li
    className="sorting__item is-active"
    key="BY_PRICE_ASC"
    onClick={[Function]}
  >
    Самый дешёвый
  </li>
  <li
    className="sorting__item"
    key="BY_SPEED_ASC"
    onClick={[Function]}
  >
    Самый быстрый
  </li>
</ul>
`;
